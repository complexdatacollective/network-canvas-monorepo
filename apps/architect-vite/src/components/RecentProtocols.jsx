import React, { Component } from 'react';
import { connect } from 'react-redux';
import { get } from 'lodash';
import { Flipped } from 'react-flip-toolkit';
import PropTypes from 'prop-types';
import ProtocolStack from './ProtocolStack';

const getRecentProtocols = (state) => get(state, 'recentProtocols', [])
  .slice(0, 4);

class RecentProtocols extends Component {
  renderRecentProtocol = (protocol) => (
    <div
      key={encodeURIComponent(protocol.filePath)}
      className="recent-protocols__protocol"
    >
      <Flipped flipId={encodeURIComponent(protocol.filePath)}>
        <ProtocolStack
          protocol={protocol}
        />
      </Flipped>
    </div>
  );

  renderWelcomeText = () => (
    <div className="recent-protocols__welcome">
      <h1>Getting Started</h1>
      <p>
        Welcome to Network Canvas Architect! To get started, use the buttons above to create a new
        interview protocol, or open an existing one from elsewhere. When you return to this screen
        later, recent protocols you have opened will be shown here.
      </p>
    </div>
  );

  renderProtocolList = (recentProtocols) => (
    <>
      <h3 className="recent-protocols__title" key="heading">Recently Opened Protocols</h3>
      <div className="recent-protocols__wrapper">
        {recentProtocols.map(this.renderRecentProtocol)}
      </div>
    </>
  );

  render() {
    const { show, recentProtocols } = this.props;

    if (!show) { return null; }

    return (
      <div className="recent-protocols">
        { (recentProtocols.length === 0)
          ? this.renderWelcomeText() : this.renderProtocolList(recentProtocols)}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  recentProtocols: getRecentProtocols(state),
});

RecentProtocols.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  recentProtocols: PropTypes.array.isRequired,
  show: PropTypes.bool,
};

RecentProtocols.defaultProps = {
  show: true,
};

export { RecentProtocols as UnconnectedRecentProtocols };

export default connect(mapStateToProps)(RecentProtocols);
