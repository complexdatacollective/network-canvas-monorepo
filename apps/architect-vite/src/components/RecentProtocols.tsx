import { Component } from "react";
import { Flipped } from "react-flip-toolkit";
import { connect } from "react-redux";
import { selectRecentProtocols, type StoredProtocol } from "~/ducks/modules/protocols";
import type { RootState } from "~/ducks/modules/root";
import ProtocolStack from "./ProtocolStack";

// Use StoredProtocol type instead
// type RecentProtocol = {
//   filePath: string;
//   lastModified: string;
//   name: string;
//   schemaVersion: string;
// };

type RecentProtocolsProps = {
	recentProtocols: StoredProtocol[];
	show?: boolean;
};

const getRecentProtocols = (state: RootState) => selectRecentProtocols(4)(state);

class RecentProtocols extends Component<RecentProtocolsProps> {
	renderRecentProtocol = (protocol: StoredProtocol) => (
		<div key={protocol.id} className="recent-protocols__protocol">
			<Flipped flipId={protocol.id}>
				<ProtocolStack protocol={protocol} />
			</Flipped>
		</div>
	);

	renderWelcomeText = () => (
		<div className="recent-protocols__welcome">
			<h1>Getting Started</h1>
			<p>
				Welcome to Network Canvas Architect! To get started, use the buttons above to create a new interview protocol,
				or open an existing one from elsewhere. When you return to this screen later, recent protocols you have opened
				will be shown here.
			</p>
		</div>
	);

	renderProtocolList = (recentProtocols: StoredProtocol[]) => (
		<>
			<h3 className="recent-protocols__title" key="heading">
				Recently Opened Protocols
			</h3>
			<div className="recent-protocols__wrapper">{recentProtocols.map(this.renderRecentProtocol)}</div>
		</>
	);

	render() {
		const { show, recentProtocols } = this.props;

		if (!show) {
			return null;
		}

		return (
			<div className="recent-protocols">
				{recentProtocols.length === 0 ? this.renderWelcomeText() : this.renderProtocolList(recentProtocols)}
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	recentProtocols: getRecentProtocols(state),
});

// Default props handled via TypeScript default parameter values

export { RecentProtocols as UnconnectedRecentProtocols };

export default connect(mapStateToProps)(RecentProtocols);
