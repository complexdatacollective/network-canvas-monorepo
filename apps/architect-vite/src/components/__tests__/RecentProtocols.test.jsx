import { describe, it, expect } from 'vitest';
import React from 'react';
import { render } from '@testing-library/react';
import { UnconnectedRecentProtocols } from '../RecentProtocols';

const mockProps = {
  recentProtocols: [],
};

describe('<RecentProtocols />', () => {
  it('can render?', () => {
    const { container } = render(<UnconnectedRecentProtocols {...mockProps} />);

    expect(container).toMatchSnapshot();
  });
});
