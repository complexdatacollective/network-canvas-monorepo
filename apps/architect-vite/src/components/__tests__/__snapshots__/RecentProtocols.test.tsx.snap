// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RecentProtocols /> > can render? 1`] = `<div />`;
